### PR TITLE
feat: implemented bind_custom_gradient_function

### LINKED ISSUE #10695: Add Gradient Functions to Paddle Backend
Add [Gradient Functions](https://www.paddlepaddle.org.cn/documentation/docs/en/api/index\_en.html) to Paddle backend

\_

>Please keep in mind that the proper way to link an issue to this list is to comment "- [ ] #issue\_number" while the issue's title only includes the name of the function you've chosen.

\_

- [x] variable
- [x] is\_variable
- [x] variable\_data
- [x] execute\_with\_gradients
- [x] value\_and\_grad
- [x] stop\_gradients
- [x] jac
- [x] grad

## Experimental

- [x] #26119

### LINKED ISSUE #26119: bind_custom_gradient_function

### PR DESCRIPTION
**Pull request by Moshood-Wale**
# PR Description
Implemented bind_custom_gradient_function

## Related Issue
#10695

Close #26119

## Checklist

- [x] Did you add a function?
- [x] Did you add the tests?
- [x] Did you run your tests and are your tests passing?
- [x] Did pre-commit not fail on any check?
- [x] Did you follow the steps we provided?

### CODE REVIEW

**Comment by MuhammedAshraf2020** [comment]
Hello, @Moshood-Wale!
I appreciate your contribution. Please inform me when you've completed your work so that I can review it with you.

**Comment by Moshood-Wale** [comment]
> Hello, @Moshood-Wale! I appreciate your contribution. Please inform me when you've completed your work so that I can review it with you.

Hi @MuhammedAshraf2020! I'm a little confused about the full details of the task. I would be glad if you could describe how I can go about completing the task.

**Review by vedpatwardhan — COMMENTED** [review]
I'll leave this with @MuhammedAshraf2020 to review as he's assigned to the PR, but basically you need to implement the `bind_custom_gradient_function` in `ivy` with the `paddle` backend so that it works in the way the other backends implementations do (where you can bind a custom function that computes gradients. More details about the same in [this](https://unify.ai/docs/ivy/overview/deep_dive/gradients.html#custom-gradient-functions) section. Thanks @Moshood-Wale @MuhammedAshraf2020 😄

**Comment by Moshood-Wale** [comment]
> I'll leave this with @MuhammedAshraf2020 to review as he's assigned to the PR, but basically you need to implement the `bind_custom_gradient_function` in `ivy` with the `paddle` backend so that it works in the way the other backends implementations do (where you can bind a custom function that computes gradients. More details about the same in [this](https://unify.ai/docs/ivy/overview/deep_dive/gradients.html#custom-gradient-functions) section. Thanks @Moshood-Wale @MuhammedAshraf2020 😄

Hi @vedpatwardhan. Thanks for the feedback. I will take a look at the task again.

**Comment by Moshood-Wale** [comment]
> Hello, @Moshood-Wale! I appreciate your contribution. Please inform me when you've completed your work so that I can review it with you.

Hi @MuhammedAshraf2020. Kindly review again. Thanks.

**Comment by vedpatwardhan** [comment]
Just removed my review request for now, feel free to request my review if you need me to weigh in on any points @MuhammedAshraf2020. Thanks @Moshood-Wale 😄

**Comment by Moshood-Wale** [comment]
Hi @MuhammedAshraf2020. Please note that my application at Ivy will be expiring in few days. I'm imploring you to kindly review my PR as soon as possible. 
Thanks.

**Review by vedpatwardhan — COMMENTED** [review]
Hey @Moshood-Wale, just went through the PR, actually this wouldn't solve the issue, we basically want the `custom_grad_fn` to be used when we're backpropagating rather than the forward pass , so we can't really call the `custom_grad_fn` in the forward pass (as done on line [74](https://github.com/unifyai/ivy/pull/26155/files#diff-6ec90a0972bbe1151211ecdaf261e10e01da2840b0214792349978f5c7e909daR74).
The reason behind it not working is that we also need to consider the previously computed gradients in the chain while backpropagating. 
[This](https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html) page is a good read to understand how that works. Basically [`tensorflow`](https://github.com/unifyai/ivy/blob/b1401d23501b870b2b9fd111d6e400cadfc3c498/ivy/functional/backends/tensorflow/experimental/gradients.py#L11), [`torch`](https://github.com/unifyai/ivy/blob/b1401d23501b870b2b9fd111d6e400cadfc3c498/ivy/functional/backends/torch/experimental/gradients.py#L10) and [`jax`](https://github.com/unifyai/ivy/blob/b1401d23501b870b2b9fd111d6e400cadfc3c498/ivy/functional/backends/jax/experimental/gradients.py#L17) have a backdoor to their APIs to tap into their gradient computation to insert custom gradient logic into it, there needs to be a similar backdoor in `paddle` for us to be able to get that working with the `paddle` backend.
Thanks 😄

**Comment by Moshood-Wale** [comment]
> Hey @Moshood-Wale, just went through the PR, actually this wouldn't solve the issue, we basically want the `custom_grad_fn` to be used when we're backpropagating rather than the forward pass , so we can't really call the `custom_grad_fn` in the forward pass (as done on line [74](https://github.com/unifyai/ivy/pull/26155/files#diff-6ec90a0972bbe1151211ecdaf261e10e01da2840b0214792349978f5c7e909daR74). The reason behind it not working is that we also need to consider the previously computed gradients in the chain while backpropagating. [This](https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html) page is a good read to understand how that works. Basically [`tensorflow`](https://github.com/unifyai/ivy/blob/b1401d23501b870b2b9fd111d6e400cadfc3c498/ivy/functional/backends/tensorflow/experimental/gradients.py#L11), [`torch`](https://github.com/unifyai/ivy/blob/b1401d23501b870b2b9fd111d6e400cadfc3c498/ivy/functional/backends/torch/experimental/gradients.py#L10) and [`jax`](https://github.com/unifyai/ivy/blob/b1401d23501b870b2b9fd111d6e400cadfc3c498/ivy/functional/backends/jax/experimental/gradients.py#L17) have a backdoor to their APIs to tap into their gradient computation to insert custom gradient logic into it, there needs to be a similar backdoor in `paddle` for us to be able to get that working with the `paddle` backend. Thanks 😄

Thanks for the review @vedpatwardhan. I will work on the task again based on your review.

**Comment by Moshood-Wale** [comment]
Hi @vedpatwardhan! Kindly review again.
Thanks.

**Review by vedpatwardhan — CHANGES_REQUESTED** [review]
Unfortunately this method wouldn't solve the problem either because we only want our `custom grad_fn` to be used when someone does a `ivy.execute_with_gradients` and not when they do the forward pass.
For e.g. this is an example we want to get working with the `paddle` backend,
```
import ivy


ivy.set_backend('torch')
x = ivy.array(50.0)
inter_func = lambda x: ivy.log1p(ivy.exp(x))

# args –> ((xs, ret), upstream)
def custom_grad_fn(*args):
    args1 = (1 - 10 / (1 + args[0][0]))
    return args[1] * args1

inter_func = ivy.bind_custom_gradient_function(
inter_func, custom_grad_fn
)
func = lambda x: ivy.sum(inter_func(x) ** 2)

ret, grad = ivy.execute_with_gradients(func, x)
print(grad)
```
I couldn't really find a clear alternative in `paddle`, but seems like [`paddle.Tensor.register_hook`](https://www.paddlepaddle.org.cn/documentation/docs/en/api/paddle/Tensor/register_hook_en.html#register-hook) could be an alternative, where we could apply our gradient function to all inputs being passed to the function during the forward pass. Ofc there's a possibility where `paddle` doesn't really have a way to add custom gradients, in which case we can't support this in our backends, but I think `paddle.Tensor.register_hook` is worth a try. Thanks @Moshood-Wale 😄

**Review by vedpatwardhan — APPROVED** [review]
lgtm! Just updated the code to get things working, basically `paddle.autograd.PyLayer` is almost the same as `torch.autograd.Function`, so a similar backend implementation to `torch` was useful to implement this function in the `paddle` backend. Feel free to merge the PR @MuhammedAshraf2020, thanks @Moshood-Wale 😄

**Comment by Moshood-Wale** [comment]
Thanks for your reviews @vedpatwardhan.
I learned a lot.

**Comment by Moshood-Wale** [comment]
> lgtm! Just updated the code to get things working, basically `paddle.autograd.PyLayer` is almost the same as `torch.autograd.Function`, so a similar backend implementation to `torch` was useful to implement this function in the `paddle` backend. Feel free to merge the PR @MuhammedAshraf2020, thanks @Moshood-Wale 😄

Hi @MuhammedAshraf2020. Kindly merge the PR.
Thanks.

**Comment by MuhammedAshraf2020** [comment]
LGTM! Thanks @Moshood-Wale
